app/src/main/java/de/danoeh/antennapod/activity/MainActivity.java

```diff
@@ -511,4 +516,61 @@ public Snackbar showSnackbarAbovePlayer(CharSequence text, int duration) {
     public Snackbar showSnackbarAbovePlayer(int text, int duration) {
         return showSnackbarAbovePlayer(getResources().getText(text), duration);
     }
+
+    //Hardware keyboard support
+    @Override
+    public boolean onKeyUp(int keyCode, KeyEvent event) {
+        Intent serviceIntent = new Intent(this, PlaybackService.class);
+        serviceIntent.putExtra(MediaButtonReceiver.EXTRA_HARDWAREBUTTON, true);
+
+        AudioManager audioManager = (AudioManager) getSystemService(AUDIO_SERVICE);
+        Integer customKeyCode = null;
+
+        switch (keyCode) {
+            case KeyEvent.KEYCODE_ENTER: //Fallthrough
+            case KeyEvent.KEYCODE_SPACE:
+                //Play/Pause
+                customKeyCode = KeyEvent.KEYCODE_MEDIA_PLAY_PAUSE;
+                break;
+            case KeyEvent.KEYCODE_DPAD_LEFT:
+                //Go Back
+                customKeyCode = KeyEvent.KEYCODE_MEDIA_REWIND;
+                break;
+            case KeyEvent.KEYCODE_DPAD_RIGHT:
+                //Go Forward
+                customKeyCode = KeyEvent.KEYCODE_MEDIA_FAST_FORWARD;
+                break;
+            case KeyEvent.KEYCODE_M:
+                //Mute/Unmute
+                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+                    audioManager.adjustStreamVolume(AudioManager.STREAM_MUSIC, AudioManager.ADJUST_TOGGLE_MUTE, 0);
+                } else {
+                    if (audioManager.getStreamVolume(AudioManager.STREAM_MUSIC) == 0) {
+                        //Unmute
+                        audioManager.setStreamVolume(AudioManager.STREAM_MUSIC, audioManager.getStreamMaxVolume(AudioManager.STREAM_MUSIC), 0);
+                    } else {
+                        //Mute
+                        audioManager.setStreamVolume(AudioManager.STREAM_MUSIC, 0, 0);
+                    }
+                }
+                break;
+            case KeyEvent.KEYCODE_DPAD_UP:
+                //Raise volume
+                audioManager.adjustStreamVolume(AudioManager.STREAM_MUSIC, AudioManager.ADJUST_RAISE, AudioManager.FLAG_SHOW_UI);
+                break;
+            case KeyEvent.KEYCODE_DPAD_DOWN:
+                //Raise volume
+                audioManager.adjustStreamVolume(AudioManager.STREAM_MUSIC, AudioManager.ADJUST_LOWER, AudioManager.FLAG_SHOW_UI);
+                break;
+            default:
+                return super.onKeyUp(keyCode, event);
+        }
+
+        if (customKeyCode != null) {
+            serviceIntent.putExtra(MediaButtonReceiver.EXTRA_KEYCODE, customKeyCode);
+            ContextCompat.startForegroundService(this, serviceIntent);
```

Comment:
The media button receiver is not a service. I would try using a broadcast instead. I think we have a method somewhere that is called "sendLocalBroadcast" or something like that.

Reply:
I am using the `PlaybackService` (l. 523), I only use the `EXTRA_KEYCODE` from `MediaButtonReceiver` as name of the extras.

Reply:
If I send the new `KeyEvent` to `MediaButtonReceiver` the delay between the press and the responding action increased (on my emulator) a lot. Furthermore I took the intent from the `MediaButtonReceiver`, so the behaviour is the same.